### PR TITLE
feature: create-project improvement

### PR DESCRIPTION
**Pull request by polunzh**
- project:create, check is package exists
- project:create, support load from local
- project:create, download starter from npm, and cache it locally
- refactor: make create project read some config from package.json
- refactor: capsulate copy file code to a function

### CODE REVIEW

File: src/cli/project/create/create.js [hunk 1]
```diff
@@ -128,87 +93,311 @@ function createDirectoryContents({ fromPath, toPath, blacklist, symbolicLinks })
...
+}
+
+function getStarter(starterDir, starterEntryPath) {
+  let configFilePath = path.join(starterDir, './starter.config.js');
+
+  // for compatible with old starter
+  if (!fs.existsSync(configFilePath)) {
+    configFilePath = path.resolve(starterDir, starterEntryPath);
+    if (!fs.existsSync(configFilePath)) {
+      throw new Error(`entry file: "${configFilePath}" is no exists`);
```
Comment (wangshijun) [inline src/cli/project/create/create.js]: `is no exists` => `does not exist`

Comment (wangshijun) [reply src/cli/project/create/create.js]: entry file => `starter config file`

File: src/cli/project/create/create.js [hunk 2]
```diff
@@ -128,87 +93,311 @@ function createDirectoryContents({ fromPath, toPath, blacklist, symbolicLinks })
...
+  // eslint-disable-next-line
+  const starter = require(configFilePath);
+
+  return starter;
+}
+
+function getStarterPackageConfig(starterPath) {
+  const packageJSONPath = path.join(starterPath, 'package.json');
+  if (!fs.existsSync(packageJSONPath)) {
+    throw new Error('no package.json file in starter');
```
Comment (wangshijun) [inline src/cli/project/create/create.js]: `no xxx in starter` => `xxx not found in starter directory ${name}`

File: src/cli/project/create/create.js [hunk 3]
```diff
@@ -128,87 +93,311 @@ function createDirectoryContents({ fromPath, toPath, blacklist, symbolicLinks })
...
+  return packageJSON;
+}
+
+function installNodeDependencies(dir, registry) {
+  printInfo('Installing dependencies...');
+
+  let command = `${pm} install --color`;
+  if (registry) {
+    command = `${command} --registry=${registry}`;
+    debug('using registry:', registry);
```
Comment (wangshijun) [inline src/cli/project/create/create.js]: highlight this output, just like `custom mirror` in `forge install`

File: src/cli/project/create/create.js [hunk 4]
```diff
@@ -128,87 +93,311 @@ function createDirectoryContents({ fromPath, toPath, blacklist, symbolicLinks })
...
+  }
+
+  return shell.exec(command, { cwd: dir });
+}
+
+async function downloadStarter(template, dest, registry = '') {
+  fs.mkdirSync(dest, { recursive: true });
+  debug('starter directory:', dest);
+
+  printInfo('Downloading package...');
```
Comment (wangshijun) [inline src/cli/project/create/create.js:189]: shall we use a spinner here?

File: src/cli/project/create/create.js [hunk 5]
```diff
@@ -128,87 +93,311 @@ function createDirectoryContents({ fromPath, toPath, blacklist, symbolicLinks })
...
+  const packageName = stdout.trim();
+  await tar.x({ file: path.join(dest, packageName), C: dest, strip: 1 });
+
+  return dest;
+}
+
+async function getTargetDir(targetDirectory) {
+  let result = targetDirectory;
+  if (fs.existsSync(result) && fs.readdirSync(result).length) {
+    printWarning(`Target directory ${result} is exists, please choose other:`);
```
Comment (wangshijun) [inline src/cli/project/create/create.js]: `is exists` => `already exists and not empty`

File: src/cli/project/create/create.js [hunk 6]
```diff
@@ -128,87 +93,311 @@ function createDirectoryContents({ fromPath, toPath, blacklist, symbolicLinks })
...
+}
+
+async function checkStarterVersion(starterName, localVersion, remoteVersion) {
+  if (semver.lt(localVersion, remoteVersion)) {
+    const { confirm } = await inquirer.prompt([
+      {
+        type: 'confirm',
+        name: 'confirm',
+        default: false,
+        message: `${starterName}: new version ${remoteVersion} is available, upgrade?`,
```
Comment (wangshijun) [inline src/cli/project/create/create.js]: shell we hightlight key information in the message

File: src/cli/project/create/create.js [hunk 7]
```diff
@@ -128,87 +93,311 @@ function createDirectoryContents({ fromPath, toPath, blacklist, symbolicLinks })
...
     if (yes || (Array.isArray(starter.questions) && starter.questions.length === 0)) {
       Object.assign(config, defaults, starter.defaults || {});
     } else {
       const answers = await inquirer.prompt(starter.questions);
       Object.assign(config, answers);
     }
 
-    // Sync files
-    printInfo('application config', config);
+    // sync files
```
Comment (wangshijun) [inline src/cli/project/create/create.js]: Capitalize first word of these lines

**Review by wangshijun — CHANGES_REQUESTED** [review]

File: src/cli/project/create/create.js (hunk 4 above)
Comment (polunzh) [reply src/cli/project/create/create.js:189]: `spinner`和下载的输出会相互干扰，打印的效果不好。

**Review by wangshijun — APPROVED** [review]

**Pull request merged**